Fix for issue #161

hargobind opened the pull request:
Fix handler for chrome.tabs.onUpdated where `tab.url` is missing.

It checks if `tab.url` is undefined then call `chrome.tabs.get(tabId)`. Then proceed handling the tab update. Since `tab.url` may still exist in some scenarios, I have moved the code that was in `onUpdated.addListener()` to its own handler function.

ossilator commented on src/js/background.js:
```diff
@@ -1712,17 +1712,13 @@ var tgs = (function() {
-        gsUtils.log(tabId, 'tab url changed. changeInfo: ', changeInfo);
-        checkForTriggerUrls(tab, changeInfo.url);
-        queueSessionTimer();
+      if (tab.url) {
```
unless you can point to specific circumstances where this would be actually needed, i don't see the point of keeping it.

hargobind replied:
Since I don't know the inner workings of TMS, I didn't want to change any code that was already there. The main purpose of my change was to add `chrome.tabs.get()`. The rest of the code is the same (just rearranged).

ossilator replied:
i know, but this kind of sloppy programming is a recipe for disaster, or at least unmaintainable code ...

(as things stand, it seems unlikely that someone else will "do a proper job".)

hargobind replied:
Normally I don't just go removing code, even when it appears to be a typo. So I had to look deeper into this to see where `changeInfo.url` was introduced, or at least how long it's been around. I searched as far back as 2017 and it's still there. The code inside the `if ()` statement has changed since then, so maybe in an old browser version `changeInfo.url` actually had a value. But it's been obsolete for a long time, so I think it's safe to remove.

Since I'm still learning GitHub, can you tell me if I can update this PR, or if I need to cancel it and add a new one?

ossilator replied:
i haven't looked deep enough to judge the argument, but at face value it seems fishy - if some code is obsolete, then you should remove it first, in a self-contained commit that clearly explains what exactly (appears) to have been going on. there is a good chance that more related stuff is obsolete, or that the logic was broken entirely.

you apparently managed to update the PR just fine. note however that it's preferable to have a clean history with atomic commits, so massage it with `git rebase -i` (squash away fixups, etc.) and `git push -f` the branch.

briantully replied:
If you haven’t looked deep enough, why offer an opinion or code review? Saying something is “sloppy” or “fishy” without offering specific and researched recommendations is just lazy IMHO.

ossilator replied:
with sufficient experience, it doesn't take a lot to recognize a job half done.
my review _is_ specific, in that it says that the argumentation isn't conclusive. that doesn't say anything about the change itself, but it should be reason enough to look at it critically.
of course i could do more myself, but we all have our priorities.

hargobind replied:
After looking more at the code around it, simply removing code inside of `if (changeInfo.url)` is not the way to go.

There are 3 calls inside of that if-statement: log(), checkForTriggerUrls(), and queueSessionTimer(). The 1st and 3rd calls are found in the other `chrome.tabs.onEVENT()` callbacks, so they should be kept here to maintain similarity. I changed the remaining call to `if (tab.url) { checkForTriggerUrls(...) }`.

@ossilator Thanks for your eye in finding this typo and starting the code review.

In the future, please keep in mind that typos do happen, especially on massive code projects that are constantly evolving. Don't be so quick to look at a bug and jump to conclusions with disparaging characterizations like: sloppy, a recipe for disaster, unmaintainable, fishy, or a job half done. It's repugnant and bad form.

ossilator replied:
why are you speaking of "typo"? there was no talk about anything that would qualify as that.

would you mind rebasing & cleaning up the branch as i suggested? that would make reviewing it properly a lot easier.

hargobind replied:
The typo is in the line above your code review -- `if (changeInfo.url) {`. Instead, it should have been `tab.url`.

hargobind replied:
I believe I got the rebasing working properly. Please have a look now.

ossilator replied:
> The typo is in the line above your code review [...]

i see. but this isn't in fact related to to my previous review.

> I believe I got the rebasing working properly. 

you did on a technical level, but the outcome isn't quite what i had hoped for.
for background, the standard i work with is https://wiki.qt.io/Commit_Policy. while TMS is obviously not bound by it, i think it's a quite reasonable set of rules (including the external pages it references).

so i think you should do this:
- `git branch for-later`
- `git reset --hard HEAD~2`
- now fix the typo, and commit it
- `git checkout -f for-later -- background.js
- commit it
- `git branch -D for-later`
- `git push -f ...`

i.e., fix the obvious first (and explain what the implications of the bug were), then proceed with the bigger fix (which contains a minor refactoring).

i still think you need to get rid of the "possibly". you need to be able to cite an authoritative reference (a spec, or a commit in chromium), as otherwise you're just poking in the dark. maybe this is in fact a bug in chromium which should be fixed by them (we can still include a workaround here, but it should be clearly designated as such).

napolux commented:
LGTM, but please rebase before this can be merged. Thanks @hargobind !!!!

hargobind commented:
I believe I got the rebasing working properly. Please have a look now.

hargobind commented:
Thanks @ossilator for your help as I become a better Git user. I appreciate standards and will rework my code/commits.

On further testing, I don't know for sure that my changes are the solution here.

I did some testing in Brave and FireFox. I found that when a tab loads a new URL, `chrome.tabs.onUpdated()` is triggered, and the `changeInfo` parameter has `status` == 'loading' and `url` has a value. So I want to understand why `changeInfo.url` doesn't have a value in Chrome before I feel confident about my proposed changes. Tomorrow I'm going to try testing with a fresh copy of Chrome with no extra extensions to see if the missing `changeInfo.url` is from an extension conflict, or if it's an issue with the recent versions of Chrome. (I'll test in Edge too.)

hargobind committed:
* Fix #161 where suspending a tab turns to default icon with 3 dots and white background

Due to conflicts with other extensions, the `tab.url` and `changeInfo.url` arguments in `chrome.tabs.onUpdated()` may be missing.
See https://bugs.chromium.org/p/chromium/issues/detail?id=1305284
If `tab.url` is undefined, get the full tab object via `chrome.tabs.get(tabId)`. Then proceed handling the tab update.
Since `tab.url` may still exist in some scenarios, the code in `chrome.tabs.onUpdated()` has been moved to its own handler function.

hargobind commented:
Thanks again @ossilator for your suggestions, both on commits and on how to rebase. I've pushed my latest changes which I simplified into a single commit.

Looking deeper into this, the original issue (#161) appears to be a [bug introduced in Chrome 99](https://bugs.chromium.org/p/chromium/issues/detail?id=1305284&q=onUpdated&can=2). The bug is the result of a conflict with multiple extensions where the `Event` object is mutable. Therefore one extension's event listener can inadvertently affect a different extension's event listener, as the (modified) `Event*` is carried over to subsequent listeners. Supposedly issue was [fixed in Chrome's code](https://chromium-review.googlesource.com/c/chromium/src/+/3531082), and is either working in the latest version, or will be when Chrome 102 is released.

In the meantime, feel free to apply this patch which I have simplified after doing extensive testing in a clean copy of Chrome (plus Brave and Edge).

Changes I made:
- I added more authoritative comments, and a reference to the Chrome bug.
- I confirmed with testing that  `changeInfo.url` should always exists when `changeInfo.status` == `loading`. So I updated that property from the retrieved tab object.

ossilator commented:
thanks, excellent research!

workaround looks fine to the degree i can judge it without becoming a chrome extension dev first.

for pedantry, see https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html

note that applying this workaround makes sense only if @gioxx actually makes a release anytime soon, given that chrome 101 will have this fixed already.

hargobind commented:
Closing this since the upstream bug in #161 has been resolved.
